### PR TITLE
Add abeja.training.Jobs/Job

### PR DESCRIPTION
**Pull request by ishikawa**
Connect abeja-inc/platform-planning#3300

- https://github.com/abeja-inc/abeja-platform-sdk/pull/20
- https://github.com/abeja-inc/abeja-platform-sdk/pull/21

### Example

```python
from abeja.training import Client, JobStatus

client = Client()
adapter = client.job_definitions()
definition = adapter.get('flower-classification')

for job in definition.jobs().list():
    if job.status == JobStatus.COMPLETE:
        print('Job {} was completed!', job.job_id)
```

```

### CODE REVIEW

File: abeja/train/instance_type.py [hunk 1]
```diff
@@ -1 +1,32 @@
-from abeja.training.instance_type import InstanceType  # noqa: F401
```
Comment (ishikawa) [inline abeja/train/instance_type.py:1]: InstanceType という enum は entity として使われてるのかと思いましたが、実際には APIClient の parameter validation としてしか使われていませんでした。

元々はこれに合わせて、instance_type や exec_env を enum/class にするつもりでしたが、文字列のままでも困らない（というか追加のときにも困らない）ので、文字列のままにするつもりです。

File: abeja/train/instance_type.py [hunk 2]
```diff
@@ -1 +1,32 @@
...
+from enum import Enum
+from abeja.exceptions import BadRequest
+
+
+# WARNING. This enum will be deprecated in the future version.
+# It is now only used in parameter validation.
+class InstanceTypeValidator(Enum):
```
Comment (ishikawa) [inline abeja/train/instance_type.py:7]: 混乱の元なので、適切な名前に変更

File: abeja/training/api/client.py [hunk 3]
```diff
@@ -772,14 +771,17 @@ def create_training_job(
             - Unauthorized: Authentication failed
             - InternalServerError
         """
+        # TODO: Remove
+        from abeja.train.instance_type import InstanceTypeValidator
```
Comment (ishikawa) [inline abeja/training/api/client.py:775]: abeja.train をトップレベルで import すると循環参照になるので一旦ここ

File: abeja/training/job_definition.py [hunk 4]
```diff
@@ -149,7 +150,7 @@ class JobDefinitionVersion():
     def __init__(self, api: APIClient,
                  organization_id: str,
                  job_definition_id: str,
-                 job_definition_version: int,
+                 job_definition_version_id: int,
```
Comment (ishikawa) [inline abeja/training/job_definition.py:153]: 今更気づいたのですが、このままだと int の `job_definition_version` と Optional[JobDefinitionVersion] の `job_definition_version` を区別できないので今のうちに名前を変えました。

File: abeja/training/job_definition.py [hunk 5]
```diff
@@ -251,6 +252,204 @@ def modified_at(self) -> str:
...
+        always returns an object filled with default values.
+        """
+        stats = (response.get('statistics') or {})
+        stages = {}
+        if 'stages' in stats:
+            stages = stats.pop('stages')
+
+        statistics = Statistics(**stats)
+        for name, values in stages.items():
+            statistics.add_stage(name=name, **values)
```
Comment (ishikawa) [inline abeja/training/job_definition.py:321]: API response から Statistics を作るコードです。以下のようなレスポンスが返ってきているので大丈夫なはず。

```json
      "statistics": {
        "stages": {
          "validation": {},
          "train": {
            "loss": 0.5,
            "debug": "debug"
          }
        },
        "progress_percentage": 0.11,
        "num_epochs": 100,
        "epoch": 11
      },
```

File: abeja/training/job_definition.py [hunk 6]
```diff
@@ -251,6 +252,211 @@ def modified_at(self) -> str:
...
+            return None
+
+        stages = {}
+        if 'stages' in response:
+            stages = response.pop('stages')
+
+        statistics = Statistics(**response)
+        for name, values in stages.items():
+            statistics.add_stage(name=name, **values)
+        return statistics
```
Comment (ishikawa) [inline abeja/training/job_definition.py:315]: API response から Statistics を作るコードです。以下のようなレスポンスが返ってきているので大丈夫なはず。

```json
      "statistics": {
        "stages": {
          "validation": {},
          "train": {
            "loss": 0.5,
            "debug": "debug"
          }
        },
        "progress_percentage": 0.11,
        "num_epochs": 100,
        "epoch": 11
      },
```

File: abeja/training/job_definition.py [hunk 7]
```diff
@@ -251,6 +252,211 @@ def modified_at(self) -> str:
...
+        return self.__job_definition
+
+    @property
+    def job_definition_version(self) -> JobDefinitionVersion:
+        """Get the job definition version of this job."""
+        if self.__job_definition_version is None:
+            self.__job_definition_version = JobDefinitionVersions(
+                api=self.__api, job_definition=self.job_definition).get(
+                    job_definition_version_id=self.job_definition_version_id)
+        return self.__job_definition_version
```
Comment (ishikawa) [inline abeja/training/job_definition.py:457]: 読みづらい... 😓

File: abeja/common/exec_env.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    - **EDGE**: An edge device
+    - **LOCAL**: train-local via ABEJA SDK/CLI
+    - **NONE**: For pre-trained models
+    - **UNKNOWN**: Unrecognized value
+    """
+    CLOUD = 'cloud'
+    EDGE = 'edge'
+    LOCAL = 'local'
+    NONE = 'none'
+    UNKNOWN = 'unknown'
```
Comment (ishikawa) [inline abeja/common/exec_env.py:18]: 将来的に未知の値が追加された場合に必要

File: abeja/training/job_definition.py [hunk 9]
```diff
@@ -251,6 +253,211 @@ def modified_at(self) -> str:
...
+        statistics = Job.build_statistics(response.get('statistics'))
+
+        return Job(
+            api=api,
+            organization_id=organization_id,
+            job_definition_id=response.get('job_definition_id', ''),
+            job_definition_version_id=response.get('job_definition_version', 0),
+            job_id=response.get('id', response.get('training_job_id', '')),
+            instance_type=response.get('instance_type', ''),
+            exec_env=ExecEnv.from_value(cast(str, response.get('exec_env'))),
```
Comment (ishikawa) [inline abeja/training/job_definition.py:338]: response の値は Any なので cast 必要

File: abeja/common/instance_type.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    def is_cpu(self):
+        return self.cpu_type == CPUType.CPU
+
+    def __str__(self):
+        if self.cpu_category:
+            s = '{}:{}-{:.2f}'.format(self.cpu_type.value, self.cpu_category.value, self.cpu)
+        else:
+            s = '{}-{:.2f}'.format(self.cpu_type.value, self.cpu)
+        return s.rstrip('.0')
```
Comment (ishikawa) [inline abeja/common/instance_type.py:68]: `gpu-1.00` みたいにならないように必要

File: abeja/train/instance_type.py (hunk 1 above)
Comment (ishikawa) [reply abeja/train/instance_type.py:1]: >元々はこれに合わせて、instance_type や exec_env を enum/class にするつもりでしたが、文字列のままでも困らない（というか追加のときにも困らない）ので、文字列のままにするつもりです。

やっぱり enum/class にします...

File: abeja/common/instance_type.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    CPU_0_25 = cast('InstanceType', None)
+    CPU_1 = cast('InstanceType', None)
+    CPU_2 = cast('InstanceType', None)
+    CPU_4 = cast('InstanceType', None)
+    CPU_8 = cast('InstanceType', None)
+    CPU_16 = cast('InstanceType', None)
+    GPU_A1 = cast('InstanceType', None)
+    GPU_B1 = cast('InstanceType', None)
+    GPU_B4 = cast('InstanceType', None)
+    GPU_B8 = cast('InstanceType', None)
```
Comment (ishikawa) [inline abeja/common/instance_type.py:79]: 1. Enum との整合性を保つため、`InstanceType.CPU_1` のように使いたい
2. しかし、そのためにはクラス定義のあとに代入が必要
3. そうすると my.py が型チェックできない
4. ので、あらかじめダミーの値 (`None`) を与えています

他にいい方法あればやめたい...。

Comment (ishikawa) [reply abeja/common/instance_type.py:79]: 更に subclass を作る、という手もある？

```python
__InstanceType = NamedTuple('InstanceType', ...)

class BaseInstanceType(__InstanceType):
    ...
    # define all methods and attributes

class InstanceType(BaseInstanceType):
    CPU_0_25 = BaseInstanceType.parse('cpu-0.25')
    ...
```

Comment (ishikawa) [reply abeja/common/instance_type.py:79]: しかし、これだと `BaseInstanceType` で型づけされてしまうので、`InstanceType` を要求するときにエラーになる 😇

Comment (ishikawa) [reply abeja/common/instance_type.py:79]: 他の案としては、InstanceType は Enum っぽく振る舞えるので、

1. 定数は module attribute にする
2. 他の Enum (exec_env, job_status) も全てこの形に変更する

にするのは個人的にいいと思った

File: abeja/common/exec_env.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
+from typing import NamedTuple
+
+
+__ExecEnv = NamedTuple('ExecEnv', [
+    ('value', str),
+])
+
+
+class ExecEnv(__ExecEnv):
```
Comment (ishikawa) [inline abeja/common/exec_env.py:9]: 敢えて Enum にしていません。参考: https://abeja.slack.com/archives/C4F9HEF0U/p1588729810368300

File: abeja/training/job_status.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
+from typing import NamedTuple
+
+
+__JobStatus = NamedTuple('InstanceType', [
+    ('value', str),
+])
+
+
+class JobStatus(__JobStatus):
```
Comment (ishikawa) [inline abeja/training/job_status.py:9]: 敢えて Enum にしていません。参考: https://abeja.slack.com/archives/C4F9HEF0U/p1588729810368300

File: abeja/training/api/client.py [hunk 14]
```diff
@@ -779,8 +780,14 @@ def create_training_job(
...
-        if instance_type is not None and InstanceType.to_enum(instance_type):
-            data['instance_type'] = instance_type
+        if instance_type is not None:
+            # validation
+            try:
+                InstanceType.parse(instance_type)
+                data['instance_type'] = instance_type
+            except ValueError:
+                error_message = "'{}' is an invalid instance_type".format(instance_type)
+                raise BadRequest(error=error_message, error_description=error_message, status_code=400)
```
Comment (ishikawa) [inline abeja/training/api/client.py:790]: instance_type の validation は新しい InstanceType を使うようにしました。ここだけ既存の挙動を変えていますが、今後新しい InstanceType が追加されたときのためにもこちらの方がいいと思います。

File: abeja/training/api/client.py [hunk 15]
```diff
@@ -1023,7 +1030,7 @@ def update_statistics(self, organization_id: str, job_definition_name: str, trai
                 statistics.add_stage(name=Statistics.STAGE_TRAIN, accuracy=0.9, loss=0.05)
                 statistics.add_stage(name=Statistics.STAGE_VALIDATION, accuracy=0.8, loss=0.1, key2=2)
 
-                response = api_client.update_statistics(statistics)
+                response = api_client.update_statistics(statistics.get_statistics())
```
Comment (ishikawa) [inline abeja/training/api/client.py:1033]: docstring の修正

File: abeja/training/job_definition.py [hunk 16]
```diff
@@ -558,58 +802,233 @@ def update(self, job_definition_version: int, description: str) -> JobDefinition
...
+                jobs.update_statistics(job_id, statistics)
+
+        Params:
+            - **job_id** (str): Job ID
+            - **statistics** (:class:`Statistics`): statistics
+        """
+        self.__api.update_statistics(organization_id=self.organization_id,
+                                     job_definition_name=self.job_definition_name,
+                                     training_job_id=job_id,
+                                     statistics=statistics.get_statistics())
```
Comment (ishikawa) [inline abeja/training/job_definition.py:1023]: abeja.train.Client では例外を無視していたりしたのですが、とりあえず余計なことをせずにそのままにしています。やっぱり無視した方がいいんですかね？

File: abeja/training/job_definition.py [hunk 17]
```diff
@@ -558,58 +802,233 @@ def update(self, job_definition_version: int, description: str) -> JobDefinition
...
 
 # Iterator classes
 
 
-class JobDefinitionIterator(SizedIterable[JobDefinition]):
+T = TypeVar('T')
+
+
+class AbstractSizedIterator(SizedIterable[T]):
```
Comment (ishikawa) [inline abeja/training/job_definition.py:1031]: ページングが必要なレスポンスが JobDefinitionVersion の他に Job も増えたので、抽象クラス化

File: doc/source/common/exec_env.rst [hunk 18]
```diff
@@ -0,0 +1,6 @@
+abeja.common.exec_env
```
Comment (ishikawa) [inline doc/source/common/exec_env.rst:1]: ドキュメントはまだ適当です... 😓

File: abeja/common/docker_image_name.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+__DockerImageName = NamedTuple('DockerImageName', [
+    ('repository', str),
+    ('name', str),
+    ('tag', Optional[str]),
+    ('host', Optional[str]),
+])
+
+
+class DockerImageName(__DockerImageName):
```
Comment (ysku) [inline abeja/common/docker_image_name.py:13]: namedtuple を継承するのをはじめてみました
初期化に関する記述量が少ないのと namedtuple のメソッドも使えるから便利ですね、勉強になりました 🙏 
https://docs.python.org/3.6/library/collections.html#namedtuple-factory-function-for-tuples-with-named-fields

読んでて見つけたのですが `__slots__ = ()` とするのは不要ですかね？

File: abeja/training/job_definition.py [hunk 20]
```diff
@@ -251,6 +273,229 @@ def modified_at(self) -> str:
...
+        self.__job_definition_version = job_definition_version
+
+    @staticmethod
+    def build_statistics(response: Optional[Dict[str, Any]]) -> Optional[Statistics]:
+        if response is None:
+            return None
+
+        stages = {}
+        if 'stages' in response:
+            stages = response.pop('stages')
```
Comment (ysku) [inline abeja/training/job_definition.py:331]: stages がない場合もあり得そう? 
https://github.com/abeja-inc/model_api/blob/6c005dbc618e4a434b019607de2b47315170c971/tests/utils/custom_type_validator_test.py#L312-L339

File: abeja/training/job_definition.py (hunk 16 above)
Comment (ysku) [reply abeja/training/job_definition.py:1023]: 途中で API の一時的な不具合があったときに、学習が止まってしまうと、 ~殺害され~ 怒られそうという背景があった気がします 😓

File: abeja/training/job_definition.py (hunk 20 above)
Comment (ishikawa) [reply abeja/training/job_definition.py:331]: あ、そうですね。テストもあるので大丈夫なはず
https://github.com/abeja-inc/abeja-platform-sdk/blob/feature/abeja-training-package-jobs/tests/training/test_job_definition.py/#L565

File: abeja/training/job_definition.py (hunk 16 above)
Comment (ishikawa) [reply abeja/training/job_definition.py:1023]: なるほどー。悩ましいけど今と同じにしようと思います。例外捕捉したければ APIClient 使えるし

**Comment by ishikawa** [comment]
あと、job_definition module に詰め込みすぎなので、次の PR で分割します。

File: abeja/training/job_definition.py (hunk 16 above)
Comment (ishikawa) [reply abeja/training/job_definition.py]: @ysku 07bc927 にて対応しました。

**Comment by ishikawa** [comment]
:azamasu:
